### PR TITLE
Move eventd enabled check from build time to runtime

### PR DESCRIPTION
**Pull request by zbud-msft**
##### Work item tracking
- Microsoft ADO **(number only)**:28728116

#### How I did it

Remove eventd enabled/slim image check from Dockerfile (build time). As part of Dockerfile expose eventd_enabled and slim image flags to ENV which will be used docker_init/start.sh to check if rsyslog plugin should be moved to rsyslog.d

#### How to verify it

Manual test/pipeline

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: slave.mk [hunk 1]
```diff
@@ -1186,6 +1186,8 @@ $(addprefix $(TARGET_PATH)/, $(DOCKER_IMAGES)) : $(TARGET_PATH)/%.gz : .platform
 			--build-arg SONIC_VERSION_CACHE=$(SONIC_VERSION_CACHE) \
 			--build-arg SONIC_VERSION_CACHE_SOURCE=$(SONIC_VERSION_CACHE_SOURCE) \
 			--build-arg image_version=$(SONIC_IMAGE_VERSION) \
+			--build-arg eventd_enabled=$(INCLUDE_SYSTEM_EVENTD) \
```
Comment (qiluo-msft) [inline slave.mk:1189]: >eventd_enabled [](http://example.com/codeflow?start=15&length=14)

let's keep using consistent words. eventd_enabled->include_eventd.

We normally use "include_" for build time flags.

File: dockers/docker-dhcp-relay/start.sh [hunk 2]
```diff
@@ -24,3 +24,8 @@ if [ $(supervisorctl status | grep -c "^dhcp-relay:") -gt 0 ]; then
     # lifetime of the process.
     /usr/bin/wait_for_intf.sh
 fi
+
+if [[ "${EVENTD_ENABLED}" == "y" ]] && [[ "${REDUCED_BUILD_SIZE}" == "n" ]]; then
```
Comment (qiluo-msft) [inline dockers/docker-dhcp-relay/start.sh:28]: >EVENTD_ENABLED [](http://example.com/codeflow?start=9&length=14)

checking one flag is enough.

File: dockers/docker-dhcp-relay/Dockerfile.j2 [hunk 3]
```diff
@@ -50,14 +54,14 @@ COPY ["files/supervisor-proc-exit-listener", "/usr/bin"]
...
-COPY ["*.json", "/etc/rsyslog.d/"]
-COPY ["files/rsyslog_plugin.conf.j2", "/etc/rsyslog.d/"]
+COPY ["*.json", "/etc/rsyslog_plugin/"]
+COPY ["files/rsyslog_plugin.conf.j2", "/etc/rsyslog_plugin/"]
 # Create dhcp_relay_regex.conf
-RUN j2 -f json /etc/rsyslog.d/rsyslog_plugin.conf.j2 /etc/rsyslog.d/events_info.json  > /etc/rsyslog.d/dhcp_relay_events.conf
-RUN rm -f /etc/rsyslog.d/rsyslog_plugin.conf.j2
-RUN rm -f /etc/rsyslog.d/events_info.json
-{% endif %}
+RUN j2 -f json /etc/rsyslog_plugin/rsyslog_plugin.conf.j2 /etc/rsyslog_plugin/events_info.json  > /etc/rsyslog_plugin/dhcp_relay_events.conf
```
Comment (qiluo-msft) [inline dockers/docker-dhcp-relay/Dockerfile.j2:63]: >j2 [](http://example.com/codeflow?start=4&length=2)

we should move j2 into start.sh, and also use include_system_eventd into the j2 itself. so no need to conditional copy inside start.sh

Comment (qiluo-msft) [reply dockers/docker-dhcp-relay/Dockerfile.j2:63]: we normally just COPY j2 files into /usr/share/sonic/templates/ inside containers.

File: slave.mk (hunk 1 above)
Comment (zbud-msft) [reply slave.mk]: Changed

File: dockers/docker-dhcp-relay/start.sh (hunk 2 above)
Comment (zbud-msft) [reply dockers/docker-dhcp-relay/start.sh]: Changed

File: dockers/docker-dhcp-relay/Dockerfile.j2 (hunk 3 above)
Comment (zbud-msft) [reply dockers/docker-dhcp-relay/Dockerfile.j2]: Done

File: dockers/docker-dhcp-relay/start.sh [hunk 4]
```diff
@@ -24,3 +24,8 @@ if [ $(supervisorctl status | grep -c "^dhcp-relay:") -gt 0 ]; then
     # lifetime of the process.
     /usr/bin/wait_for_intf.sh
 fi
+
+j2 -f json --import-env=ENVIRONMENT /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2 /usr/share/sonic/templates/rsyslog_plugin/events_info.json  > /etc/rsyslog.d/dhcp_relay_events.conf
+rm -f /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2
+rm -f /usr/share/sonic/templates/rsyslog_plugin/events_info.json
```
Comment (qiluo-msft) [inline dockers/docker-dhcp-relay/start.sh:30]: Are these 2 lines are really needed? Normally we do not remove rendering sources after rendering.

File: dockers/docker-fpm-frr/docker_init.sh [hunk 5]
```diff
@@ -112,4 +112,8 @@ TZ=$(cat /etc/timezone)
 rm -rf /etc/localtime
 ln -sf /usr/share/zoneinfo/$TZ /etc/localtime
 
+j2 -f json --import-env=ENVIRONMENT /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2 /usr/share/sonic/templates/rsyslog_plugin/events_info.json  > /etc/rsyslog.d/bgp_events.conf
+rm -f /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2
+rm -f /usr/share/sonic/templates/rsyslog_plugin/events_info.json
```
Comment (qiluo-msft) [inline dockers/docker-fpm-frr/docker_init.sh:117]: really needed?

File: dockers/docker-orchagent/docker-init.j2 [hunk 6]
```diff
@@ -74,4 +74,8 @@ TZ=$(cat /etc/timezone)
 rm -rf /etc/localtime
 ln -sf /usr/share/zoneinfo/$TZ /etc/localtime
 
+j2 -f json --import-env=ENVIRONMENT /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2 /usr/share/sonic/templates/rsyslog_plugin/events_info.json  > /etc/rsyslog.d/swss_events.conf
+rm -f /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2
+rm -f /usr/share/sonic/templates/rsyslog_plugin/events_info.json
```
Comment (qiluo-msft) [inline dockers/docker-orchagent/docker-init.j2:79]: really needed?

File: dockers/docker-dhcp-relay/start.sh (hunk 4 above)
Comment (zbud-msft) [reply dockers/docker-dhcp-relay/start.sh]: Removed

File: dockers/docker-fpm-frr/docker_init.sh (hunk 5 above)
Comment (zbud-msft) [reply dockers/docker-fpm-frr/docker_init.sh]: Removed

File: dockers/docker-orchagent/docker-init.j2 (hunk 6 above)
Comment (zbud-msft) [reply dockers/docker-orchagent/docker-init.j2]: Removed

**Review by qiluo-msft — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

File: dockers/docker-dhcp-relay/docker_init.sh [hunk 7]
```diff
@@ -21,4 +21,6 @@ chmod +x /usr/bin/wait_for_intf.sh
 # The docker container should start this script as PID 1, so now that supervisord is
 # properly configured, we exec /usr/local/bin/supervisord so that it runs as PID 1 for the
 # duration of the container's lifetime
+j2 -f json --import-env=ENVIRONMENT /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2 /usr/share/sonic/templates/rsyslog_plugin/events_info.json  > /etc/rsyslog.d/dhcp_relay_events.conf
```
Comment (qiluo-msft) [inline dockers/docker-dhcp-relay/docker_init.sh:24]: >ENVIRONMENT [](http://example.com/codeflow?start=24&length=11)

offline discussed, need to check ConfigDB flag, not env var.

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
As comments

**Comment by StormLiangMS** [comment]
hi @yxieca  could you cherry pick?

**Comment by StormLiangMS** [comment]
hi @zbud-msft  do we need this one for 202405?

**Comment by bingwang-ms** [comment]
@zbud-msft, @qiluo-msft We are only cherry-pick bug fix for 202405 branch. This seems to be an enhancement to me. Please clarify if this is a must-have for 202405 branch. Thanks!

File: files/build_templates/docker_image_ctl.j2 [hunk 8]
```diff
@@ -327,6 +327,11 @@ function postStartAction()
             docker cp $PSENSOR pmon:/usr/bin/
         fi
     fi
+{%- elif docker_container_name == "eventd" %}
+    export EVENTD_STATE=$(sonic-db-cli -s CONFIG_DB HGET 'FEATURE|eventd' 'state')
+    j2 -f json --import-env=ENVIRONMENT /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2 /usr/share/sonic/templates/rsyslog_plugin/events_info.json  > /etc/rsyslog.d/host_events.conf
+    j2 -f json --import-env=ENVIRONMENT /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2 /usr/share/sonic/templates/rsyslog_plugin/syncd_events_info.json  > /etc/rsyslog.d/syncd_events.conf
```
Comment (stepanblyschak) [inline files/build_templates/docker_image_ctl.j2:333]: @zbud-msft @qiluo-msft These commands are CPU heavy as well as those added in docker-init.j2 for multiple containers.
Any performance measurements done w.r.t fast-reboot, warm-reboot to understand the impact?
Any idea how to achieve same functionality without introducing additional CPU overhead during init?

File: dockers/docker-dhcp-relay/docker_init.sh [hunk 9]
```diff
@@ -21,4 +21,7 @@ chmod +x /usr/bin/wait_for_intf.sh
 # The docker container should start this script as PID 1, so now that supervisord is
 # properly configured, we exec /usr/local/bin/supervisord so that it runs as PID 1 for the
 # duration of the container's lifetime
+export EVENTD_STATE=$(sonic-db-cli -s CONFIG_DB HGET 'FEATURE|eventd' 'state')
+j2 -f json --import-env=ENVIRONMENT /usr/share/sonic/templates/rsyslog_plugin/rsyslog_plugin.conf.j2 /usr/share/sonic/templates/rsyslog_plugin/events_info.json  > /etc/rsyslog.d/dhcp_relay_events.conf
```
Comment (stepanblyschak) [inline dockers/docker-dhcp-relay/docker_init.sh:25]: @zbud-msft What if eventd state changes in runtime ? Is this flow unsupported ?

File: files/build_templates/docker_image_ctl.j2 [hunk 10]
```diff
@@ -327,6 +327,11 @@ function postStartAction()
...
+    systemctl restart rsyslog
```
Comment (stepanblyschak) [inline files/build_templates/docker_image_ctl.j2:334]: @zbud-msft @qiluo-msft In our testing this introduces a race between eventd's postStartAction and rsyslog-config service. rsyslog-config.service will overwrite /etc/rsyslog.conf and when at the same time we restart rsyslogd we'll get rsyslogd exited and then eventd service start failure.

Comment (stepanblyschak) [inline files/build_templates/docker_image_ctl.j2:334]: In case this fails eventd docker will be still running

Comment (stepanblyschak) [inline files/build_templates/docker_image_ctl.j2:334]: @zbud-msft @qiluo-msft We will loose syslog messages during boot up

**Review by stepanblyschak — COMMENTED** [review]
We notice few issues in 202405 regression, @zbud-msft @qiluo-msft Could you please have a look?

**Comment by stepanblyschak** [comment]
@zbud-msft @qiluo-msft Could you please have a look - https://github.com/sonic-net/sonic-buildimage/issues/20544 ?

**Pull request merged**